entities/tower.py

```diff
@@ -1,22 +1,106 @@
+import arcade
 from arcade import SpriteList
 from entities.entity import entity
 
-
 class tower(entity):
 
+    all_enemies:SpriteList = None
+
     # TODO: make filepath be data-driven. Different tower types can have a different filepath. I guess this would be filepath_data? Or something like that.
     def __init__(
         self,
-        filename: str = ":resources:images/space_shooter/playerShip1_orange.png",
+        filename: str = "images/towerImg.png",
         position: tuple[float, float] = (0, 0),
     ):
         super().__init__(filename=filename, position=position)
-
         # initialization logic goes here.
+        #Sets up when the tower last fired.
+        self.time_since_last_firing = 0.0
+        #Sets up the firing rate/
+        self.firing_rate = 2.0
+        #Sets up damage multiplier.
+        self.damage_multiplier = 1
+        #Sets up damage level.
+        self.current_tower_lv = 1
+        #Temporary resource value
+        self.resources_value = 10
+        #Value of the range a tower can shoot.
+        self.range = 50
+
 
     def on_update(self, delta_time: float = 1 / 60):
         super().on_update(delta_time)
         # implement logic
+        #Tracks rate of fire.
+
+        def towerShoot():
+            self.time_since_last_firing += delta_time
+            
+            if self.time_since_last_firing >= self.firing_rate:
+
+                self.time_since_last_firing = 0
+                #print("Bang")
+                #Bullet spawn positon to tower goes here. 
+
+        #If the resources and tower level are correct, then the tower is upgraded. If the player does not have
+        #the correct number of resources, then every value stays the same and the user is told they don't have
+        #enough resources.
+        def upgrade(resources, damage, tower_level):
+            if resources >= 30 and tower_level == 3:
```

Comment:
We don't want to hardcode level values. It should be procedural or data-driven.

Reply:
I put this in as a temporary solution to test the code until we could get data driven code in there.